libraries/llvm/rts.ll

```diff
@@ -233,29 +217,89 @@ define void @eraseNegative(%Neg %val) alwaysinline {
+    %i = phi i64 [%evidence, %entry], [%nexti, %loop]
+    %stk = load %Stk, ptr %stkp
+    %regionp = getelementptr %StkVal, %Stk %stk, i64 0, i32 2
+    %nextp = getelementptr %StkVal, %Stk %stk, i64 0, i32 3
+    %nexti = sub i64 %i, 1
+    %cmp = icmp eq i64 %nexti, 0
+
+    br i1 %cmp, label %done, label %loop
+
+done:
+    ret ptr %regionp
+
+}
+
+
+
+define { ptr, %Ref } @alloc(i64 %idx, i64 %evidence) alwaysinline {
+    %regionp = call ptr @getRegionPointer(i64 %evidence)
+    %spp = getelementptr %Region, ptr %regionp, i64 0, i64 %idx, i32 0
+    %basep = getelementptr %Region, ptr %regionp, i64 0, i64 %idx, i32 1
+    %limitp = getelementptr %Region, ptr %regionp, i64 0, i64 %idx, i32 2
+
+    %sp = load %Sp, ptr %spp
+    %base = load %Base, ptr %basep
+    %limit = load %Limit, ptr %limitp
+
+    %obj = icmp ne i64 %idx, 0
+    %size = select i1 %obj, i64 16, i64 8
+
+    %nextsp = getelementptr i8, %Sp %sp, i64 %size
+
+    %cmp = icmp ule %Sp %nextsp, %limit
+    br i1 %cmp, label %continue, label %realloc
+
+continue:
+    store %Sp %nextsp, ptr %spp
+    %intb = ptrtoint %Base %base to i64
+    %intsp = ptrtoint %Sp %sp to i64
+    %offset = sub i64 %intsp, %intb
 
-    %statenext = getelementptr %StateHeader, %State %state, i64 0, i32 0
-    %statesize = getelementptr %StateHeader, %State %state, i64 0, i32 1
+    %ret.0 = insertvalue { ptr, %Ref } undef, %Sp %sp, 0
+    %ret.1 = insertvalue { ptr, %Ref } %ret.0, %Ref %offset, 1
+    ret { ptr, %Ref } %ret.1
 
-    %regionval = load %RegionVal, %Region %region
-    %oldstate = extractvalue %RegionVal %regionval, 0
-    %oldsize = extractvalue %RegionVal %regionval, 1
+realloc:
+    %intbase = ptrtoint %Base %base to i64
+    %intlimit = ptrtoint %Limit %limit to i64
+    %arenasize = sub i64 %intlimit, %intbase
+    %empty = icmp eq i64 %arenasize, 0
+    %double = mul i64 %arenasize, 2
+    %newarenasize = select i1 %empty, i64 1024, i64 %double
 
-    %newsize = add i64 %oldsize, %size
-    %newregionval.0 = insertvalue %RegionVal undef, %State %state, 0
-    %newregionval = insertvalue %RegionVal %newregionval.0, i64 %newsize, 1
-    store %RegionVal %newregionval, %Region %region
+    %newbase = call ptr @realloc(ptr %base, i64 %newarenasize)
+    %newlimit = getelementptr i8, %Base %newbase, i64 %newarenasize
+    %newsp = getelementptr i8, %Base %newbase, i64 %arenasize
+    %newnextsp = getelementptr i8, %Sp %newsp, i64 %size
 
-    store %State %oldstate, ptr %statenext
-    store i64 %size, ptr %statesize
+    store %Base %newbase, ptr %basep
+    store %Limit %newlimit, ptr %limitp
+    store %Sp %newnextsp, ptr %spp
+
+    %ret..0 = insertvalue { ptr, %Ref } undef, %Sp %newsp, 0
+    %ret..1 = insertvalue { ptr, %Ref } %ret..0, %Ref %arenasize, 1
+    ret { ptr, %Ref } %ret..1
 
-    %ptr = getelementptr i8, ptr %state, i64 16
-    ret %Ref %ptr
 }
 
+define ptr @getPtr(%Ref %ref, i64 %idx, i64 %evidence) alwaysinline {
```

Comment:
The `%idx` is for the 3 different kinds of values, right?

Reply:
Yes, perhaps it should be renamed to `%kind` or something similar.